syntax = "proto3";

package daq;

service Daq{
    rpc StartDaq(startDaqRequest) returns (stream startDaqResponse) {}
    rpc StopDaq(stopDaqRequest) returns (commandResponse) {}
    rpc InitDaq(configDaqRequest) returns (commandResponse) {}
    rpc ConfigureDaq(configDaqRequest) returns (commandResponse) {}
    rpc GetConfiguration(getConfigRequest) returns (getConfigResponse) {}
    rpc DaqStatus(daqStatusRequest) returns (daqStatusResponse) {}
}

message CallState {
  enum State {
    LISTENING = 0;
    RECEIVING = 1;
    STOPPED = 2;
  }
  State state = 2;
}

enum ResultCode{
    OK = 0;
    STARTED = 1;
    QUEUED = 2;
    FAILED = 3;
    UNKNOWN = 4;
    REJECTED = 5;
    NOT_ALLOWED = 6;
    ABORTED = 7;
}

message CallInfo {
    string data_types_received=1;
    string files_written=2;
    string extra_info=3;
}

message startDaqRequest{
    string modes_to_start = 1;
}

message startDaqResponse {
    oneof start_daq_response {
        CallState call_state = 1;
        CallInfo call_info = 2;
    }
}

// Empty messages as no params expected but this is in case it changes.
message stopDaqRequest{
    
}

message getConfigRequest{
    
}

message getConfigResponse{
    string config = 1;
}

message daqStatusRequest{

}

message daqStatusResponse{
    string status = 1;
}

message commandResponse{
    ResultCode result_code = 1;
    string message = 2;
}

message configDaqRequest{
    string config = 1;
}
/*
message noParamRequest{

}
*/